pandas/core/groupby/categorical.py

```diff
@@ -75,21 +75,22 @@ def recode_for_groupby(
         return c, None
 
     # sort=False should order groups in as-encountered order (GH-8868)
-    cat = c.unique()
 
-    # See GH-38140 for block below
-    # exclude nan from indexer for categories
-    take_codes = cat.codes[cat.codes != -1]
-    if cat.ordered:
-        take_codes = np.sort(take_codes)
-    cat = cat.set_categories(cat.categories.take(take_codes))
-
-    # But for groupby to work, all categories should be present,
-    # including those missing from the data (GH-13179), which .unique()
-    # above dropped
-    cat = cat.add_categories(c.categories[~c.categories.isin(cat.categories)])
-
-    return c.reorder_categories(cat.categories), None
+    # xref GH:46909: Re-ordering codes faster than using (set|add|reorder)_categories
+    all_codes = np.arange(c.categories.nunique(), dtype=np.int8)
+    # GH 38140: exclude nan from indexer for categories
+    # TODO: Probably why dropa=False doesn't work GH 36327
+    unique_notnan_codes = unique1d(c.codes[c.codes != -1])
```

Comment:
I'm working on a branch that fixes `dropna=False` for categorical (currently writing tests, but at least reducers and transformers seem to be working), this section of code is unchanged in that. My thinking here is that this is determining the categories and categories cannot include nan, so dropping -1 is okay even for `dropna=False`.

Reply:
Ah okay thanks for the insights. I'll remove my comment then referencing dropna=False.

I just found it odd that the this generates `self.grouping_vector` and it specifically drops -1 codes unlike `sort=True`